### PR TITLE
Feature/issue 4 thinh

### PR DESCRIPTION
**Pull request by tcthinhsqt**

### CODE REVIEW

File: backend/app/Http/Controllers/AdminController.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Validator;
+use Illuminate\Support\Facades\Auth;
+use App\Http\Requests\AdminRequest;
+
+class AdminController extends Controller
+{
+    public function index(Request  $request) {
```
Comment (hdkhoasgt) [inline backend/app/Http/Controllers/AdminController.php:12]: @tcthinhsqt 
Please add PHP-doc for all functions

File: backend/app/Http/Controllers/AdminController.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        return view('Student.adminlogin');
+    }
+
+    public function login(AdminRequest $request) {
+        $validated = $request->only('username', 'password');
+        if (Auth::attempt($validated)) {
```
Comment (hdkhoasgt) [inline backend/app/Http/Controllers/AdminController.php:18]: @tcthinhsqt 
You can use `$request->validated()`

File: backend/app/Http/Controllers/AdminController.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            return redirect()->route('student.index');
+        } else {
+            return redirect()->back()->withErrors('Sai tên tài khoản hoặc mật khẩu!')->withInput();
```
Comment (hdkhoasgt) [inline backend/app/Http/Controllers/AdminController.php:21]: @tcthinhsqt 
Please use the lang values. 
Do not use const value in the source code. Add to lang or config.

File: backend/resources/lang/en/auth.php

File: backend/routes/web.php [hunk 4]
```diff
@@ -21,4 +21,13 @@
...
+
+
+Route::get('/admin/student', ['as' => 'student.index', 'uses' => 'StudentController@index'])
+    ->middleware('auth');
+Route::post('/admin/student', ['as' => 'student.store', 'uses' => 'StudentController@store'])
+    ->middleware('auth');
+Route::put('/admin/student', ['as' => 'student.update', 'uses' => 'StudentController@update'])
+    ->middleware('auth');
+//Route::resource('/admin/student', 'StudentController')->middleware('auth');
+
```
Comment (hdkhoasgt) [inline backend/routes/web.php]: @tcthinhsqt 
Please change to below.
```
Route::resource('/admin/students', 'StudentController')->middleware('auth');
``` 
https://laravel.com/docs/7.x/controllers#resource-controllers

File: backend/app/Http/Middleware/RedirectIfAuthenticated.php [hunk 5]
```diff
@@ -19,7 +19,7 @@ class RedirectIfAuthenticated
     public function handle($request, Closure $next, $guard = null)
     {
         if (Auth::guard($guard)->check()) {
-            return redirect(RouteServiceProvider::HOME);
+            return redirect()->route('student.index');
```
Comment (hdkhoasgt) [inline backend/app/Http/Middleware/RedirectIfAuthenticated.php:22]: @tcthinhsqt 
Please revert this change. Change the value of `RouteServiceProvider::HOME `

File: backend/app/Models/Grade.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+class Grade extends Model
+{
+    public $timestamps = false;
+    protected $fillable = [
+        'student_id',
+        'subject',
+        'grade',
+    ];
+    public function Student() {
```
Comment (hdkhoasgt) [inline backend/app/Models/Grade.php:15]: @tcthinhsqt 
Please follow naming conventions
https://github.com/alexeymezenin/laravel-best-practices#follow-laravel-naming-conventions

File: backend/app/Models/Grade.php [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        return $this->belongsTo('App\Models\Student', 'student_id', 'id');
```
Comment (hdkhoasgt) [inline backend/app/Models/Grade.php:16]: @tcthinhsqt 
```suggestion
        return $this->belongsTo(Student::class, 'student_id', 'id');
```

File: backend/app/Models/Student.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+use Illuminate\Database\Eloquent\Model;
+
+class Student extends Model
+{
+    public $timestamps = false;
+    protected $fillable = [
+        'name',
+        'birthday'
+    ];
+    public function Grade() {
```
Comment (hdkhoasgt) [inline backend/app/Models/Student.php:14]: @tcthinhsqt
Please follow naming conventions
https://github.com/alexeymezenin/laravel-best-practices#follow-laravel-naming-conventions

File: backend/app/Models/Student.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        return $this->hasMany('App\Models\Grade', 'student_id', 'id');
```
Comment (hdkhoasgt) [inline backend/app/Models/Student.php:15]: @tcthinhsqt 
```suggestion
        return $this->hasMany(Grade::class, 'student_id', 'id');
```

File: backend/config/auth.php [hunk 10]
```diff
@@ -71,6 +76,11 @@
             'model' => App\User::class,
```
Comment (hdkhoasgt) [inline backend/config/auth.php:76]: @tcthinhsqt 
Just change this class to `App\Models\Admin::class`

Revert this file

File: backend/database/seeds/AdminsSeeder.php [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+     * Run the database seeds.
+     *
+     * @return void
+     */
+    public function run()
+    {
+        DB::table('admins')->insert([
+            'username' => 'taikhoan',
+            'password' => Hash::make('matkhau'),
+            'email' =>  'taikhoan@gmail.com'
```
Comment (hdkhoasgt) [inline backend/database/seeds/AdminsSeeder.php:17]: @tcthinhsqt 
Requirement is admin/admin/admin@gmail.com

File: backend/database/seeds/DatabaseSeeder.php [hunk 12]
```diff
@@ -11,6 +11,6 @@ class DatabaseSeeder extends Seeder
      */
     public function run()
     {
-        // $this->call(UserSeeder::class);
+        $this->call(AdminsSeeder::class);
```
Comment (hdkhoasgt) [inline backend/database/seeds/DatabaseSeeder.php:14]: @tcthinhsqt 
Please add factory for Student table. Create 500 students.

File: backend/app/Http/Requests/AdminRequest.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'username' => 'bail|required|min:6',
```
Comment (hdkhoasgt) [inline backend/app/Http/Requests/AdminRequest.php:28]: @tcthinhsqt 
Do not check min length here. It's for creating and updating.

File: backend/resources/views/Layout/layout.blade.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+
+    <title>Laravel Exercise - @yield('title')</title>
+
+    <!-- Fonts -->
+    <link href="https://fonts.googleapis.com/css?family=Nunito:200,600" rel="stylesheet">
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
Comment (hdkhoasgt) [inline backend/resources/views/Layout/layout.blade.php:12]: @tcthinhsqt 
Move the js files to bottom of file

File: backend/resources/views/Layout/layout.blade.php [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    @stack('styles')
+</head>
+<body>
+@section('sidebar')
+    This is the master sidebar.
+@show
+<div class="container">
+    @yield("content");
+</div>
+@stack('js')
```
Comment (hdkhoasgt) [inline backend/resources/views/Layout/layout.blade.php:87]: @tcthinhsqt 
Move this stack to after all js imports

File: backend/resources/views/Student/adminlogin.blade.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
+<!-- Stored in resources/views/child.blade.php -->
```
Comment (hdkhoasgt) [inline backend/resources/views/Student/adminlogin.blade.php:1]: @tcthinhsqt
Please follow naming conventions
https://github.com/alexeymezenin/laravel-best-practices#follow-laravel-naming-conventions
https://laravel.com/docs/7.x/controllers#resource-controllers

File: backend/resources/views/welcome.blade.php [hunk 17]
```diff
@@ -8,7 +8,7 @@
 
         <!-- Fonts -->
         <link href="https://fonts.googleapis.com/css?family=Nunito:200,600" rel="stylesheet">
-
```
Comment (hdkhoasgt) [inline backend/resources/views/welcome.blade.php:11]: @tcthinhsqt 
Please revert this file. Do not exist in requirement

File: backend/routes/web.php [hunk 18]
```diff
@@ -15,4 +15,19 @@
...
+    ->middleware('auth');
+
+
+Route::get('/admin/student', ['as' => 'student.index', 'uses' => 'StudentController@index'])
+    ->middleware('auth');
+Route::post('/admin/student', ['as' => 'student.store', 'uses' => 'StudentController@store'])
+    ->middleware('auth');
+Route::put('/admin/student', ['as' => 'student.update', 'uses' => 'StudentController@update'])
+    ->middleware('auth');
+//Route::resource('/admin/student', 'StudentController')->middleware('auth');
```
Comment (hdkhoasgt) [inline backend/routes/web.php:32]: @tcthinhsqt 
Please remove prefix admin.

Comment (hdkhoasgt) [inline backend/routes/web.php:32]: @tcthinhsqt 
Please follow https://laravel.com/docs/7.x/controllers#resource-controllers

**Review by hdkhoasgt — CHANGES_REQUESTED** [review]
@tcthinhsqt 
Please confirm 🙇

File: backend/resources/views/welcome.blade.php (hunk 17 above)
Comment (hdkhoasgt) [reply backend/resources/views/welcome.blade.php:11]: @tcthinhsqt 
Just revert. Not remove it. Please

File: backend/routes/web.php [hunk 19]
```diff
@@ -12,7 +12,16 @@
 | contains the "web" middleware group. Now create something great!
 |
 */
-
 Route::get('/', function () {
-    return view('welcome');
+    return redirect()->route('student.index');
 });
```
Comment (hdkhoasgt) [inline backend/routes/web.php:17]: @tcthinhsqt 
Please add WelcomeController for this route

File: backend/routes/web.php [hunk 20]
```diff
@@ -12,7 +12,16 @@
...
+Route::get('/login', ['as' => 'login.index', 'uses' =>'AdminController@index'])
+    ->middleware('guest');
+Route::post('/login', ['as' => 'login', 'uses' =>'AdminController@login'])
+    ->middleware("throttle:10,2");;
+Route::get('logout', ['as' => 'logout', 'uses' =>'AdminController@logout']);
+
+Route::get('student/search', ['as' => 'student.search', 'uses' => 'StudentController@search'])
+    ->middleware('auth');
```
Comment (hdkhoasgt) [inline backend/routes/web.php:25]: @tcthinhsqt 
Please remove this route. Search by `student.index` route

File: backend/routes/web.php [hunk 21]
```diff
@@ -12,7 +12,16 @@
...
-
 Route::get('/', function () {
-    return view('welcome');
+    return redirect()->route('student.index');
 });
+Route::get('/login', ['as' => 'login.index', 'uses' =>'AdminController@index'])
+    ->middleware('guest');
+Route::post('/login', ['as' => 'login', 'uses' =>'AdminController@login'])
+    ->middleware("throttle:10,2");;
+Route::get('logout', ['as' => 'logout', 'uses' =>'AdminController@logout']);
```
Comment (hdkhoasgt) [inline backend/routes/web.php:22]: @tcthinhsqt 
Use POST for this route

File: backend/resources/views/student/test-login.blade.php [hunk 22]
```diff
@@ -0,0 +1,7 @@
+{{--@if (session('msg'))--}}
+{{--    <div class="alert alert-danger">--}}
+{{--        {{ session('msg') }}--}}
+{{--    </div>--}}
+{{--@endif--}}
+
+<p>Đăng nhập thành công</p>
```
Comment (hdkhoasgt) [inline backend/resources/views/student/test-login.blade.php:7]: @tcthinhsqt 
Please remove this file

File: backend/package-lock.json [hunk 23]
```diff
@@ -1,9949 +0,0 @@
-{
```
Comment (hdkhoasgt) [inline backend/package-lock.json:1]: @tcthinhsqt 
Please revert this file

File: backend/public/css/adminlogin.css [hunk 24]
```diff
@@ -0,0 +1,54 @@
+form {
```
Comment (hdkhoasgt) [inline backend/public/css/adminlogin.css:1]: @tcthinhsqt 
Please rename to ` backend/public/css/admin/login.css`

File: backend/public/css/createform.css [hunk 25]
```diff
@@ -0,0 +1,16 @@
+form {
```
Comment (hdkhoasgt) [inline backend/public/css/createform.css:1]: @tcthinhsqt 
Please rename to `backend/public/css/student/create.css`

File: backend/public/css/studentmanager.css [hunk 26]
```diff
@@ -0,0 +1,37 @@
+#searchdiv {
```
Comment (hdkhoasgt) [inline backend/public/css/studentmanager.css:1]: @tcthinhsqt
Please rename to backend/public/css/student/index.css

File: backend/resources/views/layout/layout.blade.php [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        .m-b-md {
+            margin-bottom: 30px;
+        }
+        .container {
+            margin-top: 5rem !important;
+            display: flex;
+            flex-flow: column wrap;
+            /*align-items: center !important;*/
```
Comment (hdkhoasgt) [inline backend/resources/views/layout/layout.blade.php:74]: @tcthinhsqt 
Please remove this comment

File: backend/resources/views/student/admin-login.blade.php [hunk 28]
```diff
@@ -0,0 +1,31 @@
+<!-- Stored in resources/views/child.blade.php -->
```
Comment (hdkhoasgt) [inline backend/resources/views/student/admin-login.blade.php:1]: @tcthinhsqt 
Please rename to `backend/resources/views/admin/login.blade.php`

Comment (hdkhoasgt) [inline backend/resources/views/student/admin-login.blade.php:1]: @tcthinhsqt 
Fix the comment

File: backend/resources/views/layout/layout.blade.php [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    </style>
+    @stack('styles')
+</head>
+<body>
+@section('sidebar')
+    This is the master sidebar.
```
Comment (hdkhoasgt) [inline backend/resources/views/layout/layout.blade.php:81]: @tcthinhsqt 
Please add to lang file

File: backend/resources/views/student/admin-login.blade.php [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+
+
+@push('styles')
+
+    <link rel="stylesheet" href="{{ asset('css/adminlogin.css') }}">
+
+@endpush
+@extends('layout.layout')
+
+@section('title', 'Admin login')
```
Comment (hdkhoasgt) [inline backend/resources/views/student/admin-login.blade.php:11]: @tcthinhsqt 
Please add to lang file `'Admin login'`

File: backend/resources/views/student/admin-login.blade.php [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+
+@section('content')
+    <form action="{{route('login')}}" method="post">
+        @csrf
+        <div>
+            <h1>Login</h1>
+        </div>
+        <input name="username" placeholder="Username" type="text" value="{{old('username')}}">
+        <input name="password" placeholder="Password" type="password">
+        <button name="submit" type="submit">LOGIN</button>
```
Comment (hdkhoasgt) [inline backend/resources/views/student/admin-login.blade.php:24]: @tcthinhsqt 
Please add to lang file `LOGIN`

File: backend/resources/views/student/create-form.blade.php [hunk 32]
```diff
@@ -0,0 +1,80 @@
+<!-- Stored in resources/views/child.blade.php -->
```
Comment (hdkhoasgt) [inline backend/resources/views/student/create-form.blade.php:1]: @tcthinhsqt 
Rename to `backend/resources/views/student/create.blade.php`

File: backend/resources/views/student/admin-login.blade.php (hunk 31 above)
Comment (hdkhoasgt) [reply backend/resources/views/student/admin-login.blade.php:24]: @tcthinhsqt 
Please add all text to lang file

File: backend/resources/views/student/edit-form.blade.php [hunk 33]
```diff
@@ -0,0 +1,92 @@
+<!-- Stored in resources/views/child.blade.php -->
```
Comment (hdkhoasgt) [inline backend/resources/views/student/edit-form.blade.php:1]: @tcthinhsqt 
Rename to `backend/resources/views/student/edit.blade.php`

File: backend/resources/views/student/student.blade.php [hunk 34]
```diff
@@ -0,0 +1,103 @@
+<!-- Stored in resources/views/child.blade.php -->
```
Comment (hdkhoasgt) [inline backend/resources/views/student/student.blade.php:1]: @tcthinhsqt 
Rename to `backend/resources/views/student/index.blade.php`

File: backend/app/Http/Requests/StoreStudentRequest.php [hunk 35]
```diff
@@ -0,0 +1,40 @@
+<?php
```
Comment (hdkhoasgt) [inline backend/app/Http/Requests/StoreStudentRequest.php:1]: @tcthinhsqt 
Rename to `backend/app/Http/Requests/Student/StoreRequest.php`

File: backend/config/auth.php [hunk 36]
```diff
@@ -66,9 +66,9 @@
     */
 
     'providers' => [
-        'users' => [
+        'admins' => [
             'driver' => 'eloquent',
-            'model' => App\User::class,
+            'model' => App\Models\Admin::class,
```
Comment (hdkhoasgt) [inline backend/config/auth.php:71]: @tcthinhsqt 
Keep only this change. revert others in this file

File: backend/database/seeds/GradesSeeder.php [hunk 37]
```diff
@@ -0,0 +1,18 @@
+<?php
```
Comment (hdkhoasgt) [inline backend/database/seeds/GradesSeeder.php:1]: @tcthinhsqt 
Please remove this file

**Review by hdkhoasgt — CHANGES_REQUESTED** [review]

File: backend/app/Http/Controllers/WelcomeController.php [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+ * Class WelcomeController
+ */
+class WelcomeController extends Controller
+{
+    /**
+     * @return RedirectResponse
+     */
+    public function index(): RedirectResponse
+    {
+        return redirect()->route('student.index');
```
Comment (hdkhoasgt) [inline backend/app/Http/Controllers/WelcomeController.php:18]: @tcthinhsqt 
Please render view welcome.blade.php

File: backend/resources/lang/en/student-form.php [hunk 39]
```diff
@@ -0,0 +1,23 @@
+<?php
```
Comment (hdkhoasgt) [inline backend/resources/lang/en/student-form.php:1]: @tcthinhsqt 
Rename to `backend/resources/lang/en/student.php `

File: backend/resources/lang/en/slide-bar.php [hunk 40]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+return [
+    // resources/lang/en/slide-bar.php
```
Comment (hdkhoasgt) [inline backend/resources/lang/en/slide-bar.php:4]: @tcthinhsqt
Rename to `backend/resources/lang/en/sidebar.php`

File: backend/resources/views/student/student.blade.php (hunk 34 above)
Comment (hdkhoasgt) [reply backend/resources/views/student/student.blade.php]: @tcthinhsqt 
Please fix this

File: backend/app/Http/Controllers/StudentController.php [hunk 41]
```diff
@@ -0,0 +1,256 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Facade\Ignition\Support\Packagist\Package;
```
Comment (hdkhoasgt) [inline backend/app/Http/Controllers/StudentController.php:5]: @tcthinhsqt 
Please remove unsed lib

**Review by hdkhoasgt — CHANGES_REQUESTED** [review]
@tcthinhsqt 
Please confirm 
Read spec carefully https://docs.google.com/spreadsheets/d/12Z4jH4arqmLzo232QT618ao0Gc-n-aSPwBXnwTMW46o/edit#gid=2035710556&range=K3

 - [x] Per page 10
 - [ ] Birthday (Y/m/d)
 - [ ] Button first and last at pagination
 - [x] GPA format (##.#)
 - [x] Column search - show desc or asc

Screen Shot 2021-09-09 at 19 09 30


Column sort example
Screen Shot 2021-09-09 at 19 16 47

File: backend/database/seeds/GradesSeeder.php (hunk 37 above)
Comment (tcthinhsqt) [reply backend/database/seeds/GradesSeeder.php]: I have already removed this file before

File: backend/resources/views/student/student.blade.php (hunk 34 above)
Comment (tcthinhsqt) [reply backend/resources/views/student/student.blade.php]: Done

File: backend/app/Http/Requests/Student/StoreRequest.php [hunk 42]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        $before_date = '2015-01-01';
+        $after_date = '1990-01-01';
```
[the remark this answers is not in the document]
Comment (tcthinhsqt) [reply backend/app/Http/Requests/Student/StoreRequest.php:29]: please grant me the permission of this document to access
https://docs.google.com/spreadsheets/d/12Z4jH4arqmLzo232QT618ao0Gc-n-aSPwBXnwTMW46o/edit#gid=2035710556&range=K3

**Review by hdkhoasgt — CHANGES_REQUESTED** [review]
@tcthinhsqt 
Please fix 
 - Birthday (Y/m/d)
 - Button first and last at pagination
 - GPA format (##.#) (04.0 -> 4.0) -> do not add leading zero
 - Column search - show desc or asc (Does not keep order when click search or refresh)
Screen Shot 2021-09-10 at 19 07 37

File: backend/app/Http/Requests/Student/StoreRequest.php [hunk 43]
```diff
@@ -0,0 +1,40 @@
...
+        return [
+            'name' => 'bail|required|bail',
+            'birthday' => 'bail|required|date|before:'
+                . $before_date . '|after:'
+                . $after_date,
+            'math' => 'bail|required|integer|between:1,10',
+            'music' => 'bail|required|integer|between:1,10',
+            'english' => 'bail|required|integer|between:1,10',
+        ];
+    }
```
[the remark this answers is not in the document]
Comment (hdkhoasgt) [reply backend/app/Http/Requests/Student/StoreRequest.php]: @tcthinhsqt 
```
        return [
            'name'     => 'bail|required',
            'birthday' => 'bail|required|date|before:'.$maxBirthday.'|after:'.$minBirthday,
            'math'     => 'bail|required|integer|between:0,10',
            'music'    => 'bail|required|integer|between:0,10',
            'english'  => 'bail|required|integer|between:0,10',
        ];
```

File: backend/database/factories/StudentFactory.php [hunk 44]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+/** @var \Illuminate\Database\Eloquent\Factory $factory */
+
+use App\Models\Student;
+use Faker\Generator as Faker;
+
+$factory->define(Student::class, function (Faker $faker) {
+    return [
+        'name' => substr($faker->firstName . ' ' . $faker->lastName,0, 20),
```
[the remark this answers is not in the document]
Comment (hdkhoasgt) [reply backend/database/factories/StudentFactory.php]: @tcthinhsqt 
Please format

**Review by hdkhoasgt — CHANGES_REQUESTED** [review]

**Pull request closed without merging**